onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc

```diff
@@ -79,27 +83,16 @@ Status ReorderPastState(
   // [B, N, head_size / x, max_length, x], where x = 16 / sizeof(T)
   int64_t chunk_size = static_cast<int64_t>(16 / past_state.DataType()->Size());
 
-  std::vector<size_t> permutation_vector = {0, 1, 3, 2, 4};
-  gsl::span<size_t> permutation(permutation_vector.data(), 5);
-
-  // "Fake" the shapes of the input and output tensors of the Transpose operation to suit our need
-  size_t offset = packed_past ? 1 : 0;
-  TensorShape transpose_input_shape_override = {past_state_shape[offset],
-                                                past_state_shape[offset + 1],
-                                                past_state_shape[offset + 2],
-                                                past_state_shape[offset + 3] / chunk_size,
-                                                chunk_size};
-
-  TensorShape transpose_output_shape_override = {past_state_shape[offset], past_state_shape[offset + 1],
-                                                 past_state_shape[offset + 3] / chunk_size, past_state_shape[offset + 2],
-                                                 chunk_size};
-
-  // TODO(hasesh): Explore perf tuning for this Transpose operation
-  return onnxruntime::cuda::Transpose::DoTranspose(*static_cast<const cudaDeviceProp*>(cuda_device_prop), cuda_stream,
-                                                   cublas_handle, permutation,
-                                                   past_state_staging, past_state,
-                                                   &transpose_input_shape_override,
-                                                   &transpose_output_shape_override);
+  cuda::ReorderPastStatesKernelLauncher(past_state.MutableDataRaw(),
```

Comment:
Possibly a further enhancement (maybe for another PR) is to try and "fuse" this operation for all layers as opposed to doing staging memcpy + reorder layer by layer in a loop. For this, it needs some other changes - like allocating one flat buffer of past states for all layers (today this is not the case) + allocate one flat "staging" buffer for all layers (today we only allocate staging buffer for one layer and re-use for all layers) and then involking just 1 memcpy and 1 re-order kernel. This should save all the kernel launch overheads.

Reply:
Actually this approach(https://github.com/microsoft/onnxruntime/tree/wangye/transpose_opt) was tried in the first place but saw performance drop. The kernel launch overheads can be ignored in the case of large input size.

Reply:
Another viable approach is to use "ping-pang" buffer and hide memcpy time. However it only works(show gain) for small inputs thus not included in this PR.